pandas/_libs/tslibs/vectorized.pyx

```diff
@@ -287,39 +245,20 @@ cpdef ndarray[int64_t] normalize_i8_timestamps(const int64_t[:] stamps, tzinfo t
     cdef:
         Py_ssize_t i, n = len(stamps)
         int64_t[:] result = np.empty(n, dtype=np.int64)
-        ndarray[int64_t] trans
-        int64_t[:] deltas
-        str typ
-        Py_ssize_t[:] pos
-        int64_t local_val, delta = NPY_NAT
-        bint use_utc = False, use_tzlocal = False, use_fixed = False
-
-    if is_utc(tz) or tz is None:
-        use_utc = True
-    elif is_tzlocal(tz):
-        use_tzlocal = True
-    else:
-        trans, deltas, typ = get_dst_info(tz)
-        if typ not in ["pytz", "dateutil"]:
-            # static/fixed; in this case we know that len(delta) == 1
-            use_fixed = True
-            delta = deltas[0]
-        else:
-            pos = trans.searchsorted(stamps, side="right") - 1
+        ndarray[intp_t] pos_
+        intp_t* pos
+        int64_t local_val
+        Localizer info = Localizer(tz)
+
+    pos_ = info.prepare(stamps)
+    pos = <intp_t*>cnp.PyArray_DATA(pos_)
 
     for i in range(n):
         if stamps[i] == NPY_NAT:
             result[i] = NPY_NAT
             continue
 
-        if use_utc:
-            local_val = stamps[i]
-        elif use_tzlocal:
-            local_val = tz_convert_utc_to_tzlocal(stamps[i], tz)
-        elif use_fixed:
-            local_val = stamps[i] + delta
-        else:
-            local_val = stamps[i] + deltas[pos[i]]
+        local_val = info.utc_val_to_local_val(stamps[i], pos, i)
```

Comment:
@scoder any ideas on how to track down the 2x performance hit here?

Reply:
`inline` in foreign `.pyx` files does not help across modules. Not sure if `@cython.final` would, but probably not either. Calling a C function instead of a method may make a difference, though. That could also live in a `.pxd` file – although I'm not sure if I ever tried to call a cimported function ( `tz_convert_utc_to_tzlocal` above) from an `inline` function in a `.pxd` file. Might work.

Definitely try changing the method to a function here, even if you then just pass the `Localizer` object into it.

Reply:
making tz_convert_utc_to_tzlocal into a function (for experimenting, put everything in vectorized.pyx rule out pyx/pxd effects) doesn't seem to have helped.

Some things i've poked at:

adding `@cython.initializedcheck(False)` in different places seems to cost a few microseconds, which i expected to go the other direction.

Keeping the L315-L322 inline but replacing each of the `if use_foo:` checks with `if info.use_foo:` appears perf-neutral.  replacing `deltas[pos[i]]` with `info.deltas[pos[i]]` caused a minor slowdown.

writing tz_convert_utc_to_tzlocal as a module-level function but pasing each of the Localizer attributes directly instead of the instance (so the body of tz_convert_utc_to_tzlocal would be identical to the current in-loop code) did not help.

Reply:
Not sure what kind of code Cython is generating with the pointer indexing but have you tried declaring a contiguous memory view for pos instead? 

https://docs.cython.org/en/stable/src/userguide/numpy_tutorial.html#declaring-the-numpy-arrays-as-contiguous

Reply:
for `pos`?  yes.